eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala

```diff
@@ -792,6 +792,8 @@ object Helpers {
           case u: UpdateFailHtlc => u.id
           case u: UpdateFailMalformedHtlc => u.id
         }.toSet
+        // these htlcs have been signed by our peer, but we haven't received their revocation and relayed them yet
+        val nonRelayedIncomingHtlcs: Set[Long] = commitment.changes.remoteChanges.signed.collect { case add: UpdateAddHtlc => add.id }.toSet
```

Comment:
Analysis:

An incoming htlc always follows this signature path:
1. receive `UpdateAddHtlc`: add htlc to `remoteChanges.proposed`
2. receive `CommitSig`, send `RevokeAndAck`: move htlc from `remoteChanges.proposed` -> `remoteChanges.acked`
3. send `CommitSig`: move htlc from `remoteChanges.acked` -> `remoteChanges.signed`
4. receive `RevokeAndAck`: remove htlc from `remoteChanges.signed`

This is why we only relay an incoming htlc when we receive remote's revocation:
https://github.com/ACINQ/eclair/blob/835b33b2b8ab6e39ff87fff652b880b19a5a2c8f/eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala#L580

However, starting from step 2., the htlc is already in our local commitment.

So, we should also ignore htlcs that are in `remoteChanges.acked`. In fact, we could take all htlcs in `remoteChanges`, which may be conceptually simpler:

```suggestion
        val nonRelayedIncomingHtlcs: Set[Long] = commitment.changes.remoteChanges.all.collect { case add: UpdateAddHtlc => add.id }.toSet
```

Reply:
Good point, I agree! This is harder to test though, because we immediately do step 2 and 3 without an opportunity to stop the channel FSM in between, but the existing test is probably already good enough.

Done in https://github.com/ACINQ/eclair/pull/2672/commits/782c8f85381558f9f94020fd347d83ae8e734f1c